internal/core/items/component_container.rs

```diff
@@ -37,12 +55,116 @@ pub struct ComponentContainer {
+                let window_item =
+                    crate::items::ItemRef::downcast_pin::<crate::items::WindowItem>(root_item)
+                        .unwrap();
+
+                // Calculate new size for both myself and the embedded window:
+                let new_width = limit_to_constraints(
+                    window_item.width(),
+                    component.as_ref().layout_info(Orientation::Horizontal),
+                );
+                let new_height = limit_to_constraints(
+                    window_item.height(),
+                    component.as_ref().layout_info(Orientation::Vertical),
+                );
+
+                Property::link_two_way(
+                    ComponentContainer::FIELD_OFFSETS.width.apply_pin(self),
+                    super::WindowItem::FIELD_OFFSETS.width.apply_pin(window_item),
+                );
+                Property::link_two_way(
+                    ComponentContainer::FIELD_OFFSETS.height.apply_pin(self),
+                    super::WindowItem::FIELD_OFFSETS.height.apply_pin(window_item),
+                );
+
+                ComponentContainer::FIELD_OFFSETS.width.apply_pin(self).set(new_width);
+                ComponentContainer::FIELD_OFFSETS.height.apply_pin(self).set(new_height);
+            } else {
+                ComponentContainer::FIELD_OFFSETS.width.apply_pin(self).set(Default::default());
+                ComponentContainer::FIELD_OFFSETS.height.apply_pin(self).set(Default::default());
+            }
+
+            self.component.replace(product);
+        }
+    }
+
+    pub fn subtree_range(self: Pin<&Self>) -> IndexRange {
+        IndexRange { start: 0, end: if self.component.borrow().is_some() { 1 } else { 0 } }
+    }
+
+    pub fn subtree_component(self: Pin<&Self>) -> ComponentWeak {
+        let rc = self.component.borrow().clone();
+        vtable::VRc::downgrade(rc.as_ref().unwrap())
+    }
+
+    pub fn visit_children_item(
+        self: Pin<&Self>,
+        index: isize,
+        order: TraversalOrder,
+        visitor: vtable::VRefMut<ItemVisitorVTable>,
+    ) -> VisitChildrenResult {
+        let rc = self.component.borrow().clone();
+        if let Some(rc) = &rc {
+            vtable::VRc::borrow_pin(rc).as_ref().visit_children_item(index, order, visitor)
+        } else {
+            VisitChildrenResult::CONTINUE
+        }
+    }
 }
 
 impl Item for ComponentContainer {
-    fn init(self: Pin<&Self>) {}
+    fn init(self: Pin<&Self>, my_component: &ComponentWeak, my_item_index: usize) {
+        self.my_component.set(my_component.clone()).or(Err(())).unwrap();
+
+        let rc = vtable::VWeak::upgrade(my_component).unwrap();
+
+        // Find my embedding item_tree_index:
+        let pin_rc = vtable::VRc::borrow_pin(&rc);
+        let item_tree = pin_rc.as_ref().get_item_tree();
+        let my_node = item_tree.iter().find(|n| matches!(n, ItemTreeNode::Item { item_array_index, .. } if (*item_array_index as usize) == my_item_index)).unwrap();
```

Comment:
Perhaps it would be better if the caller of init iterated over the item tree, instead of the items array and then requiring a linear search here for each component container.